src/app/jscaip/MCTS.ts

```diff
@@ -0,0 +1,206 @@
+import { ArrayUtils } from '../utils/ArrayUtils';
+import { MGPFallible } from '../utils/MGPFallible';
+import { MGPOptional } from '../utils/MGPOptional';
+import { Debug, Utils } from '../utils/utils';
+import { GameState } from './GameState';
+import { GameStatus } from './GameStatus';
+import { AI, AITimeLimitOptions, GameNode, MoveGenerator } from './MGPNode';
+import { Move } from './Move';
+import { Player } from './Player';
+import { Rules } from './Rules';
+
+type NodeAndPath<M extends Move, S extends GameState> = {
+    node: GameNode<M, S>,
+    path: GameNode<M, S>[],
+}
+
+/*
+ * Implement Monte-Carlo Tree Search
+ * Useful resources to understand MCTS:
+ *   - https://www.analyticsvidhya.com/blog/2019/01/monte-carlo-tree-search-introduction-algorithm-deepmind-alphago/
+ *   - https://vgarciasc.github.io/mcts-viz/
+ */
+export class MCTS<M extends Move, S extends GameState, L = void> implements AI<M, S, AITimeLimitOptions> {
+    // The exploration parameter influences the MCTS results.
+    // It is chosen "empirically" (this is the "generally recommended value" from Wikipedia)
+    public static EXPLORATION_PARAMETER: number = Math.sqrt(2);
+
+    // The longest a game can be before we decide to stop simulating it
+    public static MAX_GAME_LENGTH: number = 1000;
```

Comment:
might be interresting to make this configurable to see what it changes.

Reply:
ouais je pense que c'est intéressant de tester différentes valeurs pour trouver une optimale, mais pas suffisamment pour permettre aux utilisateurs de jouer là dessus ? Pour ça que c'est en constante et pas en option

Reply:
Configurable d'un jeu à l'autre !
Aussi, un remix de MCTS qui utiliserais à la fin, sur le noeuf final calculé, la board value ... Ouais ce serait p-e un remix du minimax du coup, mais qui choisi au hasard et pas tout pour pouvoir aller plus loin :D ? à voir :D

Reply:
Configurable d'un jeu à l'autre implique que ça doit pas être une statique, donc je change.
Sinon oui y'a pas mal de remix intéressants à tester !